dask/dataframe/core.py

```diff
@@ -3343,14 +3344,23 @@ def cummin(self, axis=None, skipna=True, out=None):
             out=out,
         )
 
+    def _condition_as_frame(self, cond):
+        if isinstance(cond, (list, np.ndarray)):
+            cond = dd.from_pandas(
+                pd.DataFrame(cond, columns=self.columns), npartitions=self.npartitions
+            ).set_index(self.index)
+        return cond
+
     @derived_from(pd.DataFrame)
     def where(self, cond, other=np.nan):
         # cond and other may be dask instance,
         # passing map_partitions via keyword will not be aligned
+        cond = self._condition_as_frame(cond)
         return map_partitions(M.where, self, cond, other, enforce_metadata=False)
 
     @derived_from(pd.DataFrame)
     def mask(self, cond, other=np.nan):
+        cond = self._condition_as_frame(cond)
```

Comment:
This seems like a nice solution, but I wonder if we can just use `elemwise` to make sure partition-alignment behavior is consistent with other row-wise operations like `assign`.

For example, instead of using `map_partitions`, maybe we could just convert `list` input to a pandas series, and then return `elemwise(M.mask, self, cond, other, meta=self._meta)`?

Reply:
`elemwise` is another possibility, I'm just not sure if it's any better. For alignment to happen, input still needs to be either a `_Frame`,  or a Dask `Array`. The condition can be 2D array, too. If it's an `Array`, then we need to give it a chunk size, which needs computing:

```python
        if isinstance(cond, (list, np.ndarray)):
            partition_size = len(self.get_partition(0))  # compute happens here
            cond = da.from_array(cond, chunks=partition_size)
            return elemwise(M.mask, self, cond, other=other, meta=self._meta)
        else:
            return map_partitions(M.mask, self, cond, other, enforce_metadata=False)
```

Or, we could require the user to provide a Dask Array, not a regular `list` or `ndarray`, and then they'd do the chunking.

What do you think is the best option?

Reply:
Since `obj.mask` technically requires `cond` to be the same shape as `obj`, it seems perfectly reasonable for dask to expect `cond` to be an object that it knows how to align with `obj` (i.e. it should be a collection or a pandas object).  I agree that it's always nice to match the pandas API, but it seems better to diverge with pandas in cases like this (e.g. `assign`).

Reply:
Have you tried converting list to `pandas` Series and call `elemwise`? I tried in this branch, but it didn't work. The index of the pandas `DataFrame` still needs to match, and inside `mask`, we only have access to dask `Index`, which would have to be computed.

Reply:
Yes, I agree. Even though we can align it ourselves, the way I did it here is hacky, and probably not a good idea to be passing around large non-lazy arrays anyway.

Reply:
True, this PR doesn't cover all the cases, but I don't want to resort to `self.index.compute()`, and there is no other way to ensure that the data is indexed the same way. With this PR, we would at least point users in the right direction if they are trying to provide an array.

Reply:
I see. What about checking the number of divisions and the value of those `divisions` in the case when a Dask collection is passed? That's a light check, and should invalidate also when that object passed is not aligned. I mean, being stricter than what you are now and raising an error if that happens.